### ISSUE TITLE
Invalid Size-Prefixed Entity in ZeroMQ API with a BlockMarker Subscription

### ISSUE THREAD

**Issue by Alexhuszagh** [nemtech/catapult-server#62]
When we look at the code to publish a block header in the ZeroMQ API, it is as follows:

# Bug

```cpp
	void ZeroMqEntityPublisher::publishBlockHeader(const model::BlockElement& blockElement) {
		auto pMessageGroup = std::make_unique<MessageGroup>(CreateHeightMessageGenerator("block header", blockElement.Block.Height));

		zmq::multipart_t multipart;
		auto marker = BlockMarker::Block_Marker;
		multipart.addmem(&marker, sizeof(marker));
		multipart.addmem(static_cast<const void*>(&blockElement.Block), sizeof(model::BlockHeader));
		multipart.addmem(static_cast<const void*>(&blockElement.EntityHash), Hash256::Size);
		multipart.addmem(static_cast<const void*>(&blockElement.GenerationHash), Hash256::Size);
		pMessageGroup->add(std::move(multipart));
		m_pSynchronizedPublisher->queue(std::move(pMessageGroup));
	}
```

When we contrast this to, let's say, the local block storage, we get the following:

```cpp
	void WriteBlockElement(const model::BlockElement& blockElement, OutputStream& outputStream) {
		// 1. write constant size data
		outputStream.write({ reinterpret_cast<const uint8_t*>(&blockElement.Block), blockElement.Block.Size });
		outputStream.write(blockElement.EntityHash);
		outputStream.write(blockElement.GenerationHash);

		// 2. write transaction hashes
		WriteTransactionHashes(outputStream, blockElement.Transactions);

		// 3. write sub cache merkle roots
		WriteSubCacheMerkleRoots(outputStream, blockElement.SubCacheMerkleRoots);
	}
```

The bug is subtle, but if we have any transaction in that block, the size-prefix of the entity will be too long for the data written in the ZeroMQ API.

# Example

A sample block reproducing the error on the symbol network.

```python
import binascii
import struct

block = binascii.unhexlify('e101000000000000f065efcd1a9a4ac749558b2f97852295baf7a9c9f4777d353456778b7813d3dc831b9973a0e890e7d06d3c5a5d1c871b0315c7314176918ff99fe52a0525090e4badbbb46f335534c8f2da93cdcea7c74452188d6a8e322720af30a6fb42dabf00000000019843819ab901000000000097e5bf260200000000a0724e18090000333fbc686b10c902ac9f4bf2b8e0c23242228898753308da997f807fefd8bcef40df5f4da93cbf7999cac4ad8b9bd98fc4228a58981e1ba829e599f47fcf8aa657a75593c1edda09ac3dfcefd72847edc8a3c17ad03331c069e7348415f99b4be89558325da0b8263039dba8fd0ed5b041697377327d433ea67f51827f44bff13d7fd669db90879c3af7eea98ea8c24460e2f6c0aaf71111c61e5f303949004b1116000000000000')
# size = 481
size = struct.unpack('<Q', block[:8])[0]
# length = 304
length = len(block)
# height = 113050
height = struct.unpack('<Q', block[112:120])[0]
```

As you can see, we are missing 177 bytes of data from the size-prefixed entity. A quick confirmation with the REST API shows these missing bytes are, in fact, a missing transaction.

```bash
$ curl localhost:3000/block/113050/transactions
[{"meta":{"height":"113050","hash":"40DF5F4DA93CBF7999CAC4AD8B9BD98FC4228A58981E1BA829E599F47FCF8AA6","merkleComponentHash":"40DF5F4DA93CBF7999CAC4AD8B9BD98FC4228A58981E1BA829E599F47FCF8AA6","index":0,"id":"5E5AA07DFEB16D4F065BFF3B"},"transaction":{"signature":"966A9F4D0D8C0F09CC81FC313FF9A4C4EA3EA7753B920E1F0ECCBB99E685D230046CB861C8664E4A2F139507DC12B81E3C85AC87828EFAE1641A59A96A5FBB0B","signerPublicKey":"63C439805416C0A2F2C3EA062094E993C5DFACC29393539E9F7DACAC543391E2","version":1,"network":152,"type":16724,"maxFee":"1000000","deadline":"9240338950","recipientAddress":"98846F170BF8B1CA8505A849322A5329CDF8EFE504002763E7","message":{"type":0,"payload":""},"mosaics":[{"id":"51A99028058245A8","amount":"899000000"}]}}]
```

# Potential Fixes

1). Change the size prefix to only include the size of the block header itself, and the entity/generation hash.
2). Include the transactions (likely less than optimal, since the ZeroMQ API also provides transactions listeners).

**Comment by BloodyRookie** [comment]
Imo this is not a bug but the whitepaper is inaccurate.
When a new block is published, onlly the block header, the entity hash and the generation hash is published. The block header is not altered and thus, the size field does not reflect the size of the data but the size of the entire block. The data size is always the same, i.e. sizeof(block header) + 2 * sizeof(hash).

We should update the whitepaper to reflect that.

**Comment by Alexhuszagh** [comment]
@BloodyRookie If that's the case though, then why have a size-prefix to begin with? I understand only the block header is published, but that really doesn't reflect the size of the entity. If the size is constant, then the size really doesn't need to be published. If the size should be published, shouldn't it reflect the size of the published data, like the rest of the published data?

**Comment by BloodyRookie** [comment]
The servers in the network distribute the block as a whole, i.e. block header with attached transactions. For the server to check if the transferred data structure is valid, the size of the data must match the size field of the structure.
When publishing via ZMQ, the block is split into the block header and the transactions, each being published as a single message. The size field in the block header is not altered in order to reflect the real data.

**Comment by Jaguar0625** [comment]
This has been addressed in version 0.9.6.3 of the whitepaper.
